File: python/recursions_backtrackings/day13/m_Coloring.py
def isSafe(node, graph, color, col):
    for adj in graph[node]:
        if color[adj] == col:
            return False
    
    return True


def solve(node, graph, color, n, m):
    if node == n:
        return True
    
    for col in range(1, m + 1):
        if isSafe(node, graph, color, col):
            color[node] = col

            if solve(node + 1, graph, color, n, m):
                return True
            
            color[node] = 0
    
    return False




# ---------------- MAIN ----------------

n = int(input("Enter number of vertices: "))
e = int(input("Enter number of edges: "))

graph = [[] for _ in range(n)]

print("Enter edges (u v):")
for _ in range(e):
    u, v = map(int, input().split())

    u -= 1
    v -= 1

    graph[u].append(v)
    graph[v].append(u)

m = int(input("Enter number of colors: "))

color = [0] * n

if solve(0, graph, color, n, m):
    print("Coloring possible!")
    for i in range(n):
        print(f"Vertex {i+1} -> Color {color[i]}")
    
else:
    print("Coloring NOT possible!")

File: java/recursions_backtrackings/Day13/M_Coloring.java
package recursions_backtrackings.Day13;
import java.util.*;

public class M_Coloring {
    private static boolean isSafe(int node, List<Integer>[] G, int[] color, int n, int col) {
        for (int it : G[node]) {
            if (color[it] == col)
                return false;
        }
        return true;
    }

    private static boolean solve(int node, List<Integer>[] G, int[] color, int n, int m) {
        if (node == n)
            return true;

        for (int i = 1; i <= m; i++) {
            if (isSafe(node, G, color, n, i)) {
                color[node] = i;

                if (solve(node + 1, G, color, n, m) == true)
                    return true;

                color[node] = 0;
            }
        }
        return false;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter number of vertices: ");
        int n = sc.nextInt();

        System.out.print("Enter number of edges: ");
        int e = sc.nextInt();


        List<Integer>[] G = new ArrayList[n];
        for (int i = 0; i < n; i++)
            G[i] = new ArrayList<>();
        
        
        System.out.println("Enter edges (u v): ");

        for (int i = 0; i < e; i++) {
            int u = sc.nextInt() - 1;
            int v = sc.nextInt() - 1;

            G[u].add(v);
            G[v].add(u);
        }


        System.out.print("Enter number of color (m): ");
        int m = sc.nextInt();

        int[] color = new int[n];

        if (solve(0, G, color, n, m)) {
            System.out.println("Coloring possible!");
            System.out.println("Assigned Colors: ");

            for(int i = 0; i < n; i++) 
                System.out.println("Vertex " + i + " -> Color " + color[i]);
        }
        else {
            System.out.print("Coloring NOT possible!");
        }

    }
}